public/app/features/serviceaccounts/ServiceAccountTable.tsx

```diff
@@ -0,0 +1,225 @@
+import React, { useMemo } from 'react';
+import Skeleton from 'react-loading-skeleton';
+
+import {
+  Avatar,
+  CellProps,
+  Column,
+  InteractiveTable,
+  Pagination,
+  Stack,
+  TextLink,
+  Button,
+  IconButton,
+  Icon,
+} from '@grafana/ui';
+import { UserRolePicker } from 'app/core/components/RolePicker/UserRolePicker';
+import { contextSrv } from 'app/core/core';
+import { AccessControlAction, OrgRole, Role, ServiceAccountDTO } from 'app/types';
+
+import { OrgRolePicker } from '../admin/OrgRolePicker';
+
+type Cell<T extends keyof ServiceAccountDTO = keyof ServiceAccountDTO> = CellProps<
+  ServiceAccountDTO,
+  ServiceAccountDTO[T]
+>;
+
+interface ServiceAccountTableProps {
+  services: ServiceAccountDTO[];
+  onRoleChange: (role: OrgRole, serviceAccount: ServiceAccountDTO) => void;
+  roleOptions: Role[];
+  onRemoveButtonClick: (serviceAccount: ServiceAccountDTO) => void;
+  onDisable: (serviceAccount: ServiceAccountDTO) => void;
+  onEnable: (serviceAccount: ServiceAccountDTO) => void;
+  onAddTokenClick: (serviceAccount: ServiceAccountDTO) => void;
+  showPaging?: boolean;
+  totalPages: number;
+  onChangePage: (page: number) => void;
+  currentPage: number;
+  isLoading: boolean;
+}
+
+export const ServiceAccountTable = ({
+  services,
+  onRoleChange,
+  roleOptions,
+  onRemoveButtonClick,
+  onDisable,
+  onEnable,
+  onAddTokenClick,
+  showPaging,
+  totalPages,
+  onChangePage,
+  currentPage,
+  isLoading,
+}: ServiceAccountTableProps) => {
+  const displayRolePicker =
+    contextSrv.hasPermission(AccessControlAction.ActionRolesList) &&
+    contextSrv.hasPermission(AccessControlAction.ActionUserRolesList);
+
+  const getCellContent = (
+    value: string,
+    original: ServiceAccountDTO,
+    isLoading: boolean,
+    columnName?: Column<ServiceAccountDTO>['id']
+  ) => {
+    if (isLoading) {
+      return columnName === 'avatarUrl' ? <Skeleton circle width={24} height={24} /> : <Skeleton width={100} />;
+    }
+    const href = `/org/serviceaccounts/${original.id}`;
+    const ariaLabel = `Edit service account's ${name} details`;
+    switch (columnName) {
+      case 'avatarUrl':
+        return (
+          <a aria-label={ariaLabel} href={href}>
+            <Avatar src={value} alt={'User avatar'} />
+          </a>
+        );
+      case 'id':
+        return (
+          <TextLink href={href} aria-label={ariaLabel} color="secondary">
+            {original.login}
+          </TextLink>
+        );
+      case 'tokens':
+        return (
+          <Stack alignItems="center">
+            <Icon name="key-skeleton-alt"></Icon>
+            <TextLink href={href} aria-label={ariaLabel} color="primary">
+              {value || 'No tokens'}
+            </TextLink>
+          </Stack>
+        );
+      default:
+        return (
+          <TextLink href={href} aria-label={ariaLabel} color="primary">
+            {value}
+          </TextLink>
+        );
+    }
+  };
```

Comment:
Does this need to be inside the component body? I was initially thinking that each cell type would have a separate function, but I think this works from a readability point of view, especially with the loading state.